### ISSUE TITLE
ModelCheckpoint filename unable to use metrics that contain a slash

### ISSUE THREAD

**Issue by its-dron** [PyTorchLightning/pytorch-lightning#4012]
## 🐛 Bug

ModelCheckpoint is unable to save filenames that reference a metric with a slash in their name. I use grouped metrics for tensorboard, and would like to save my files containing my loss: `val/loss`. However, ModelCheckpoint uses `os.path.split`, which splits the file name: https://github.com/PyTorchLightning/pytorch-lightning/blob/6ac0958166c66ed599c96737b587232b7a33d89e/pytorch_lightning/callbacks/model_checkpoint.py#L258

If I try to use 
```python
ModelCheckpoint("root/dir/{epoch}_{val/loss:.5f}")
```

The above evaluates to 
```python
self.dirpath = "root/dir/{epoch}_{val" 
self.filename = "loss:.5f}"
```

This inevitably causes failure when attempting to format the output path.

### To Reproduce

As above, log a metric with a slash, then use it in model checkpoint output

#### Code sample

```python
class Module(pl.LightningModule):
...
    def validation_step(self, batch, batch_idx):
        x, y = batch
        logits = self.forward(x)
        loss = self.loss_fn(logits, y)
        self.log('val/loss', loss, on_epoch=True)
        return loss

...
def main():
    trainer = pl.Trainer(checkpoint_callback=ModelCheckpoint("{epoch}_{val/loss:.5f}"))
```

### Expected behavior

Split only along file path boundaries, ignoring variable names yet-to-be-formatted.
Per the previous example, we'd expect:
```python
self.dirpath = "root/dir" 
self.filename = "{epoch}_{val/loss:.5f}"
```

### Environment

* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.10.0
	- tqdm:              4.50.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.5
	- version:           #1 SMP Fri Sep 4 14:19:36 UTC 2020

**Comment by ndrplz** [comment]
Cool, I was actually here to open this issue :)

I also encountered this problem. I think it's quite common to use the forward slash to group metrics in tensorboard, but this obviously clashes with this formatting.

I also tried to escape it, but with little success.

**Comment by edenlightning** [comment]
Thanks for raising the issue! would you like submit a PR for the fix?

**Comment by ozen** [comment]
#4213 doesn't fix this. Although the callback can now parse the metric name containing a slash, slashes in the resulting file name create directories during saving for obvious reasons. Since file names cannot contain slashes, the way the callback formats the file name must be changed.

**Comment by its-dron** [comment]
@ozen Am I correct that the issue now is that checkpoint names automatically include the metric name in addition to the value? 
Could we make that configurable, so I can enable/disable automatic name insertion?

**Comment by ozen** [comment]
@its-dron exactly. Making it configurable could be ugly though. Maybe the slashes can be converted to something like underscores.

**Comment by its-dron** [comment]
Personally, I'd prefer to not have the auto-inserted metric names at all. I'd prefer to have fine-grain control over what my files are named instead of having the library decide for me.

But that seems like a drastic change to drop at this point.

**Comment by ozen** [comment]
@rohitgr7 could you reopen this issue?

**Comment by rohitgr7** [comment]
Is it still an issue or not working as expected?

**Comment by ozen** [comment]
@rohitgr7 it's an issue. The way `ModelCheckpoint._format_checkpoint_name()` method is implemented prevents using metrics that contain a slash.

**Comment by rohitgr7** [comment]
Interesting, I actually added a test for this
https://github.com/PyTorchLightning/pytorch-lightning/blob/b459fd26ac773484e4c97c12e4bab221bb1609b0/tests/checkpointing/test_model_checkpoint.py#L225-L229
 not sure why is it not working. Can you put an example here, what's not working from your side. I'll have a look :)

**Comment by ozen** [comment]
`assert ckpt_name == 'epoch=4_val/loss=0.03000.ckpt'` will pass. 
But this won't create a checkpoint file named "epoch=4_val/loss=0.03000.ckpt"; it will create a directory named "epoch=4_val" and a file named "loss=0.03000.ckpt" inside that directory. If there are more than one metric with slashes, more directories...
This is the obvious result of using a slash in the name. But the way file name formatting works i.e. adding the metric name to the file name leaves the user no way to prevent this.

**Comment by rohitgr7** [comment]
I'm still little confused. What do you want your filename to be like? Don't you want to inject the metric values inside it?

**Comment by ozen** [comment]
> 
> 
> I'm still little confused. What do you want your filename to be like? Don't you want to inject the metric values inside it?

We're not talking about metric values. It injects metric **names**, which may contain a slash, which creates **directories** instead of a file. I'm sorry I'm repeating the same for the 3rd time but I don't know how else I can explain this.

**Comment by rohitgr7** [comment]
Yeah I got that directory issue, AFAIK this is expected behavior when you have slashes in the filepath.

> But the way file name formatting works i.e. adding the metric name to the file name leaves the user no way to prevent this.

your this statement got me confused.


What do you suggest, how should this work if filename contains slashes? raise a warning/exception or some replacement with another character or something else?

**Comment by its-dron** [comment]
I think there are two workable solutions:

1. Replace slashes with a safe character like an underscore or hyphen, as ozen suggested. Raising a warning is an option, but I'd find it annoying.
2. Remove or make the metric name insertion optional. Give the developer precise control over exactly how their checkpoint is named.

**Comment by edenlightning** [comment]
Feel free to send a PR with your suggestion!

**Comment by EliaCereda** [comment]
I'm also interested in a solution to this problem.

I think f-strings handle this neatly: by default `f'{metric}'` just outputs the metric's value, while `f'{metric=}'` prints both the name and the value. But I imagine this would be an unacceptable breaking change at this point.

@its-dron's option 1 would be fine for my purposes, since the existing syntax just works without any changes and I don't really care about the specific names of my checkpoints anyway. 

As a general solution, option 2 sounds better as it gives the freedom to completely customize the file names. Are you suggesting a single argument on the constructor of `ModelCheckpoint` to disable all metric names? Or do you have something in mind to specify it metric by metric?

**Comment by berebid** [comment]
@its-dron
I would prefer option 2.

@EliaCereda 
A single argument (e.g. `auto_insert_metric_name`, `prepend_metric_names` or `filepath_with_names`) on the constructor of `ModelCheckpoint` sounds good. 

If it's false, the user should be able to choose between `f'{metric}'` and `f'{metric=}'`, which is actually quite nice, and could easily be specified metric by metric.

**Comment by KrzysztofPrzygodzki** [comment]
I'm aware that a workaround, I would like propose, is absolutely not acceptable as a solution. But for a matter of pushing my project forward, I have duplicated self.log and kept it silent.

```
        self.log("validation/loss_avg", val_loss, prog_bar=True, logger=True)
        self.log("val_loss_avg", val_loss, prog_bar=False, logger=False) 
```

**Comment by mees** [comment]
I can confirm this bug. I also use tensorboard for loggin and have therefore a `self.log('val/accuracy', val_acc)` at the end of my `validation_epoch_end`.
I use these parameters for ModelCheckpoint 
```
save_top_k: 3
monitor: val/accuracy
dirpath: saved_models/
filename: '{epoch}_{val/accuracy:.4f}'
```
and a directory called `epoch=0_val` is created and a checkpoint inside with the name `accuracy=0.0000.ckpt`
I would like the checkpoint to be named  `epoch=0_val_accuracy=0.0000.ckpt` and to be placed inside the specified dirpath in this case. How can I solve this? I am using lightning 1.0.5

**Comment by maxfrei750** [comment]
Can we reopen this issue?

I think a pragmatic solution would be to just convert the slash into underscores, like @its-dron already proposed. If we can agree on a solution, then I would be happy to implement it.

**Comment by AwesomeLemon** [comment]
I think the option 2 from @its-dron is nicer because it's more flexible: it allows me to format the file name as I like, and its only cost is an additional flag in the constructor. 

Also, it's a more general solution: as mentioned in https://github.com/PyTorchLightning/pytorch-lightning/issues/4718 , the issue also occurs with the equal signs (i.e. neptune doesn't like them in file names).

**Comment by dasayan05** [comment]
So what is the solution to this issue at the moment ? I am encountering the same that @mees described, i.e. its creating folder if I use `{valid/loss}` like syntax.

**Comment by carmocca** [comment]
> So what is the solution to this issue at the moment ? I am encountering the same that @mees described, i.e. its creating folder if I use `{valid/loss}` like syntax.

See #6277

**Comment by talregev** [comment]
Can we open this issue again?
I have filename  '{epoch}-{step}-{val/f1:.2f}' and I want it will create
a filename epoch=0-step=0-val_f1=1.00 
When I choose `auto_insert_metric_name=True`, It will automatic replace all the `/` to `_` in the filename. Because in definition, filename cannot contain '/' (Slash).
```Python
auto_insert_metric_name = True
```

**Comment by carmocca** [comment]
@talregev you can do

```python
ModelCheckpoint(filename='epoch={epoch}-step={step}-val_f1={val/f1:.2f}', auto_insert_metric_name=False)
```

We aren't looking to make this replacement as you suggest since it would break backwards compatibility.

**Comment by talregev** [comment]
Can you think on other idea?
Can we add new state with new flag that will support this replacement?

**Comment by carmocca** [comment]
What is the problem with the previous suggestion?

**Comment by talregev** [comment]
Sorry I missed your suggestion. I will try it and let you know!

**Comment by talregev** [comment]
I think you need to add to docs and init documentation this example for other people.

**Comment by carmocca** [comment]
It is described here: https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint

**Comment by carmocca** [comment]
Thanks for the suggestion! Would you be interested in opening a PR? The text can be modified here https://github.com/PyTorchLightning/pytorch-lightning/blob/998265adfbe27b839e6b37407c714eb04de4a0c7/pytorch_lightning/callbacks/model_checkpoint.py#L102-L105

**Comment by manavkulshrestha** [comment]
Why is it that having the name of the checkpoint file be "epoch=n-val/f1=0.ff.ckp" as the default behavior not an option? I can't think of a reason why I would want the desired behavior to be a checkpoint file named "f1=0.ff.ckp" inside of the folder "epoch=n-val" -- do people using this library depend on this feature/behavior?

**Comment by maxfrei750** [comment]
@manavkulshrestha I'm not sure, if I understood you correctly, but a file name can never contain `/`, since the operating system wouldn't allow it. It would know if it is a file, or a file inside a folder. I hope this answers your question.
